### PR TITLE
Feature/fetch contacts from phone

### PR DESCRIPTION
**Pull request by MrrMurzik**
MA-5: when app starts first time, by clicking on the button "My Contacts" in "MyProfileActivity", appears "MyContactActivity" with a list of hardcoded users, and shows up standard system dialog with requesting permission for fetching contacts. I want that system dialog appear before creating of activity

### CODE REVIEW

**Comment by 1937Carrie** [comment]
photo_2023-02-26_12-29-37
photo_2023-02-26_12-29-37 (2)

File: app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt [hunk 1]
```diff
@@ -11,20 +14,43 @@ import androidx.recyclerview.widget.ItemTouchHelper
...
 import com.google.android.material.snackbar.Snackbar
 
 
 class MyContactsActivity : AppCompatActivity() {
 
     private lateinit var binding: ActivityMyContactsBinding
-    private val viewModel: ContactsViewModel by viewModels()
+    private var isCheckedFetching = false
+
+    private val viewModel: ContactsViewModel by viewModels {
```
Comment (1937Carrie) [inline app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt:45]: May be you should better initialize viewmodel in one of this approach https://stackoverflow.com/a/54313573/20937549 or can try https://github.com/MrrMurzik/messengerAndroid/pull/6#issuecomment-1445324380.

File: app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt [hunk 2]
```diff
@@ -125,32 +171,47 @@ class MyContactsActivity : AppCompatActivity() {
...
             .setMessage(R.string.default_deletion_dialog_message)
             .setPositiveButton(R.string.action_confirmed, listener)
-            .setNegativeButton(R.string.action_cancelled, listener)
-            .create()
-            .show()
+            .setNegativeButton(R.string.action_cancelled, listener).create().show()
     }
 
+
+    fun fetchContacts(): MutableList<User> {
```
Comment (1937Carrie) [inline app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt]: Should move the fetching of contacts to another class.

File: app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt [hunk 3]
```diff
@@ -57,6 +82,28 @@ class MyContactsActivity : AppCompatActivity() {
...
     }
 
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        when (requestCode) {
+            READ_CONTACTS_REQUEST_CODE -> {
+                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
```
Comment (1937Carrie) [inline app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt]: Call of methods in if statement and else statement is the same.

File: app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt [hunk 4]
```diff
@@ -57,6 +82,28 @@ class MyContactsActivity : AppCompatActivity() {
...
+                    setObservers()
+                } else {
+                    addSwipeToDeleteFeature()
+                    setListeners()
+                    setObservers()
+                }
+                return
+            }
+        }
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
```
Comment (1937Carrie) [inline app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt]: Usually the call to the super is written at the beginning.

File: app/src/main/java/com/example/messengerAndroid/ui/myProfile/MyProfileActivity.kt [hunk 5]
```diff
@@ -38,6 +38,8 @@ class MyProfileActivity : AppCompatActivity() {
     private fun setListeners() {
         binding.buttonContacts.setOnClickListener {
             val intent = Intent(this, MyContactsActivity::class.java)
+            val isFetchedContact = binding.checkBoxFetchContacts?.isChecked
+            intent.putExtra("Fetching key", isFetchedContact)
```
Comment (1937Carrie) [inline app/src/main/java/com/example/messengerAndroid/ui/myProfile/MyProfileActivity.kt]: Should extract the literal into constants.

**Review by 1937Carrie — CHANGES_REQUESTED** [review]
Describe in detail in the discord what you want to ask about permissions.

File: app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt (hunk 1 above)
Comment (MrrMurzik) [reply app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt:45]: Why do you think my approach is bad?

Comment (EugeneOrlov204) [reply app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt:45]: Why did you remove by viewModels in the first place? As for me it is the best choice to initialize a view model

File: app/src/main/java/com/example/messengerAndroid/ui/customView/DismissableKeyboardConstraintLayout.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+class DismissableKeyboardConstraintLayout (context: Context, attrs: AttributeSet) :
+    ConstraintLayout(context, attrs) {
+
+    override fun dispatchTouchEvent(ev: MotionEvent?): Boolean {
+        val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
+        imm.hideSoftInputFromWindow(windowToken, 0)
+        return super.dispatchTouchEvent(ev)
+    }
+}
```
Comment (EugeneOrlov204) [inline app/src/main/java/com/example/messengerAndroid/ui/customView/DismissableKeyboardConstraintLayout.kt:17]: fun View.hideKeyboard() {
    val inputMethodManager =
        context.getSystemService(android.app.Activity.INPUT_METHOD_SERVICE) as InputMethodManager
    inputMethodManager.hideSoftInputFromWindow(windowToken, 0)
}

File: app/src/main/res/layout/activity_my_profile.xml [hunk 7]
```diff
@@ -165,8 +165,23 @@
...
+        android:id="@+id/checkBoxFetchContacts"
+        android:layout_width="wrap_content"
+        android:layout_height="@dimen/checkbox_height"
+        android:layout_marginStart="@dimen/margin_for_input"
+        android:checked="false"
+        android:paddingStart="@dimen/littlePadding"
+        android:text="fetch phone contacts"
+        android:textSize="@dimen/smallFontSize"
+        app:layout_constraintBottom_toBottomOf="parent"
+        android:layout_marginBottom="10dp"
```
Comment (EugeneOrlov204) [inline app/src/main/res/layout/activity_my_profile.xml]: hardcoded

**Review by EugeneOrlov204 — APPROVED** [review]

File: app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt (hunk 1 above)
Comment (MrrMurzik) [reply app/src/main/java/com/example/messengerAndroid/ui/myContacts/MyContactsActivity.kt:45]: We have already discussed it with Eugene, I did not delete by viewModels, just add factory for viewModel to pass there arguments

File: app/src/main/java/com/example/messengerAndroid/ui/myContacts/contactsViewModel/factory/ContactsViewModelFactory.kt [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package com.example.messengerAndroid.ui.myContacts.contactsViewModel.factory
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import com.example.messengerAndroid.ui.myContacts.contactsViewModel.ContactsViewModel
+import com.example.messengerAndroid.ui.myContacts.contactsViewModel.UsersDataSelector
+
+@Suppress("UNCHECKED_CAST")
```
Comment (1937Carrie) [inline app/src/main/java/com/example/messengerAndroid/ui/myContacts/contactsViewModel/factory/ContactsViewModelFactory.kt:8]: Annotation of suppress should be removed

File: app/src/main/java/com/example/messengerAndroid/ui/myContacts/ContactFetcher.kt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+            ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
+            null,
+            null,
+            null,
+            null
+        )
+        cursor?.use {
+            while (it.moveToNext()) {
+                val id =
+                    it.getLong(it.getColumnIndexOrThrow(ContactsContract.CommonDataKinds.Phone.CONTACT_ID))
```
Comment (1937Carrie) [inline app/src/main/java/com/example/messengerAndroid/ui/myContacts/ContactFetcher.kt:21]: "or Throw" doesn't look confident

**Review by 1937Carrie — APPROVED** [review]
My review is directed at not-so-convincing places of code that can throw exceptions in specific situations. In general, the code is normal.

**Pull request merged**